lib/ansible/galaxy/collection/galaxy_api_proxy.py

```diff
@@ -85,13 +88,19 @@ def get_collection_versions(self, requirement):
         # type: (Requirement) -> Iterable[Tuple[str, GalaxyAPI]]
         """Get a set of unique versions for FQCN on Galaxy servers."""
         if requirement.is_concrete_artifact:
-            return {
-                (
-                    self._concrete_art_mgr.
-                    get_direct_collection_version(requirement),
-                    requirement.src,
-                ),
-            }
+            version = self._concrete_art_mgr.get_direct_collection_version(requirement)
+
+            err = f"Invalid version found for requirement '{requirement}': {version} ({type(version)}). "
+            err += "The version should be a semantic version string (or '*')."
+            if not isinstance(version, string_types):
```

Comment:
It could be valuable for possible future debugging to add a code comment enumerating the "wrong" things that we expect to be causing this. Is it `null`? Anything else?

Reply:
I feel like the code already expresses this more succinctly: the issue is anything that is not a string. We use `startswith` when looking for matching candidates: https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/dependency_resolution/providers.py#L278-L283
So yes, null, floats, lists, dicts would all cause an AttributeError... If it's being used in the dep resolver it should be a semantic string version or "*". I could comment a link to the cause in providers.py, but it will get out of sync.

Prior to 2.11, an invalid version, like a float, was "fixed" to be "*". In triage we discussed this, and decided we shouldn't reinstate the previous behavior because it was overly magical and would now apply to directories too.

Reply:
Yes, but the code here has a local context. You read it as "we don't want the variable to contain anything other than a string". But the underlying problem of how it is even possible that a non-string value would get assigned into this variable is not communicated by the if-clause — its semantics does not convey the deeper motivation of having this check which is an implementation detail of the YAML parser ultimately influencing this behavior. This information exists in a different place, on a different abstraction layer and a code comment could be a bridge between the UX of the end-users having to understand that `1.0` is not `"1.0"` and this check that a code reader would see.

Reply:
I still don't understand how to comment to make it clear. Feel free to suggest a comment.

Reply:
Something like this?
```suggestion
            # NOTE: The known cases causing the version to be a non-string object come from
            # NOTE: the differences in how the YAML parser normalizes ambiguous values and
            # NOTE: how the end-users sometimes expect them to be parsed. Unless the users
            # NOTE: explicitly use the double quotes of one of the multiline string syntaxes
            # NOTE: in the collection metadata file, PyYAML will parse a value containing
            # NOTE: two dot-separated integers as `float`, a single integer as `int`, and 3+
            # NOTE: integers as a `str`. In some cases, they may also use an empty value
            # NOTE: which is normalized as `null` and turned into `None` in the Python-land.
            # NOTE: Another known mistake is setting a minor part of the SemVer notation
            # NOTE: skipping the "patch" bit like "1.0" which is assumed non-compliant even
            # NOTE: after the conversion to string.
            if not isinstance(version, string_types):
```

Reply:
Merged for your future debugging pleasure.